specs/interop.md

```diff
@@ -0,0 +1,396 @@
+
+```solidity
+function hashCrossChainMessage(CrossChainMessage memory _msg) pure returns (bytes32) {
+    return keccak256(abi.encode(_msg.nonce, _msg.sender, _msg.target, _msg.value, _msg.timestamp, _msg.chainid, _msg.data));
+}
+```
+
+### CrossL2Inbox
+
+Address: `0x4200000000000000000000000000000000000022`
+
+The `CrossL2Inbox` is responsible for relaying a cross chain message on the destination chain.
+It is permissionless to finalize a cross chain message on behalf of any user. Certain protocol
+enforced invariants must be preserved to ensure safety of the protocol.
+
+#### Invariants
+
+- The timestamp of relaying message MUST be greater than or equal to the timestamp of the initiating message
+- The chain id of the initiating message MUST be in the dependency set
+- The relaying message MUST be initiated by an externally owned account such that the top level EVM call frame enters the `CrossL2Inbox`
+
+##### Only EOA Invariant
+
+The `onlyEOA` invariant on relaying a cross chain message enables static analysis on relaying messages.
+This allows for the derivation pipeline and block builders to reject relaying messages that do not
+have a corresponding initiating message.
+
+##### Transferring Ether in a Cross Chain Message
+
+The `CrossL2Inbox` MUST be initially set in state with an ether balance of `type(uint256).max`.
+There is no reason for a user to send [value](https://github.com/ethereum/execution-specs/blob/1fed0c0074f9d6aab3861057e1924411948dc50b/src/ethereum/frontier/fork_types.py#L53)
+to the `CrossL2Inbox` so there is no need to worry about overflows. This initial balance exists to
+provide liquidity for cross chain transfers of ether in a completely backwards compatible way.
+The `CrossL2Inbox` MUST only transfer out ether if the caller is the `CrossL2Outbox` from a destination
+chain.
+
+#### Relaying Messages
+
+All of the information required to satisfy the invariants MUST be included to or committed to in the calldata
+of the function that is used to execute relaying messages. Both the block builder and the smart contract use
+this information to ensure that all system invariants are held.
+
+```solidity
+function relayMessage(bytes32 _hash, CrossChainMessage memory _msg) public {
+    require(msg.sender == tx.origin);
+    require(_hash == hashCrossChainMessage(_msg));
+    require(_msg.timestamp >= block.timestamp);
+    require(L1Block.isInInteropSet(_msg.chainid));
+
+    // rest of implementation
+}
+```
+
+By including the hash in the calldata, it makes static analysis much easier for block builders.
+Without the check onchain, the block builder would need to hash the `CrossChainMessage` to build the
+`eth_getLogs` query. With the hash onchain, the block builder can directly trust the hash.
+
+### L1Block
+
+The `L1Block` contract is updated to include the set of allowed chains. The L1 Atrributes transaction
+sets the set of allowed chains. The `L1Block` contract MUST provide a public getter to check is a particular
```

Comment:
```suggestion
sets the set of allowed chains. The `L1Block` contract MUST provide a public getter to check if a particular
```

Reply:
What if different chains have the same chain ID? For example, chain ID 42069 which is a well-known chain ID for getting started OP stack deployment.

Reply:
The spec states the following:

```
The dependency set is defined by a set of chain ids. Since it is impossible to enforce uniqueness of chain ids,
social consensus MUST be used to determine the chain that represents the canonical chain id. This
particularly impacts the block builder as they SHOULD use the chain id to assist in validation
of executing messages.
```